internal/cluster/kube/servicesinformer.go

```diff
@@ -0,0 +1,167 @@
+package kube
+
+import (
+	"errors"
+	"fmt"
+	"github.com/rs/zerolog/log"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/client-go/informers"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/cache"
+	"kube-proxless/internal/cluster"
+	"kube-proxless/internal/logger"
+	"strconv"
+	"time"
+)
+
+func runServicesInformer(
+	clientSet kubernetes.Interface,
+	namespace string,
+	informerResyncInterval int,
+	upsertStore func(id, name, port, deployName, namespace string, domains []string) error,
+	deleteRouteFromStore func(id string) error,
+) {
+	opts := make([]informers.SharedInformerOption, 0)
+	if namespace != "" {
+		opts = append(opts, informers.WithNamespace(namespace))
+	}
+	informer := informers.
+		NewSharedInformerFactoryWithOptions(clientSet, time.Duration(informerResyncInterval)*time.Second, opts...).
+		Core().V1().Services().Informer()
+
+	eventHandler := cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			svc, err := parseService(obj)
+
+			if err != nil {
+				logger.Errorf(err, "Cannot process service in AddFunc handler")
+				return
+			}
+
+			addServiceToStore(clientSet, svc, upsertStore)
+
+			return
+		},
+		UpdateFunc: func(oldObj, newObj interface{}) {
+			oldSvc, err := parseService(oldObj)
+
+			if err != nil {
+				logger.Errorf(err, "Cannot process service in UpdateFunc handler")
+			}
+
+			newSvc, err := parseService(newObj)
+
+			if err != nil {
+				logger.Errorf(err, "Cannot process service in UpdateFunc handler")
+			}
+
+			updateServiceInStore(clientSet, oldSvc, newSvc, upsertStore, deleteRouteFromStore)
+			return
+		},
+		DeleteFunc: func(obj interface{}) {
+			svc, err := parseService(obj)
+
+			if err != nil {
+				logger.Errorf(err, "Cannot process service in DeleteFunc handler")
+				return
+			}
+
+			removeServiceFromStore(clientSet, svc, deleteRouteFromStore)
+
+			return
+		},
+	}
+	informer.AddEventHandler(eventHandler)
+
+	stopCh := make(chan struct{})
+	defer close(stopCh)
+	informer.Run(stopCh)
+}
+
+func parseService(obj interface{}) (*corev1.Service, error) {
+	svc, ok := obj.(*corev1.Service)
+	if !ok {
+		return nil, errors.New(fmt.Sprintf("event for invalid object; got %T want *core.Service", obj))
+	}
+	return svc, nil
+}
+
+func getPortFromServicePorts(ports []corev1.ServicePort) string {
+	port := ports[0] // TODO add possibility to manage multiple ports
+	return strconv.Itoa(int(port.TargetPort.IntVal))
+}
+
+func addServiceToStore(
+	clientset kubernetes.Interface, svc *corev1.Service,
+	upsertStore func(id, name, port, deployName, namespace string, domains []string) error,
+) {
+	if isAnnotationsProxlessCompatible(svc.ObjectMeta) {
+		deployName := svc.Annotations[cluster.AnnotationServiceDeployKey]
+
+		_, err := labelDeployment(clientset, deployName, svc.Namespace)
+
+		if err != nil {
+			logger.Errorf(err, "Error labelling deployment %s.%s", deployName, svc.Namespace)
+			// do not return here - we don't wanna break the proxy forwarding
```

Comment:
comment - will you relabel it in the future?

Reply:
Yep during the resync of the informer